src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,351 @@
+
+    /// <summary>
+    /// Handlers of non-document-related messages, indexed by handler message name.
+    /// </summary>
+    private readonly Dictionary<string, IExtensionWorspaceMessageHandlerWrapper> _handlers = new();
+
+    private readonly IExtensionMessageHandlerFactory _customMessageHandlerFactory;
+
+    // Used to protect access to _extensions, _handlers and _documentHandlers.
+    private readonly object _lockObject = new();
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public ExtensionMessageHandlerService(IExtensionMessageHandlerFactory customMessageHandlerFactory)
+    {
+        _customMessageHandlerFactory = customMessageHandlerFactory;
+    }
+
+    public ValueTask<RegisterExtensionResponse> RegisterExtensionAsync(
+        Solution solution,
+        string assemblyFilePath,
+        CancellationToken cancellationToken)
+    {
+        var assemblyFileName = Path.GetFileName(assemblyFilePath);
+        var assemblyFolderPath = Path.GetDirectoryName(assemblyFilePath)
+            ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+
+        var analyzerAssemblyLoaderProvider = solution.Services.GetRequiredService<IAnalyzerAssemblyLoaderProvider>();
+
+        CustomMessageHandlerExtension? extension;
+        lock (_lockObject)
+        {
+            // Check if the assembly is already loaded.
+            if (!_extensions.TryGetValue(assemblyFolderPath, out extension))
+            {
+                var analyzerAssemblyLoader = analyzerAssemblyLoaderProvider.CreateNewShadowCopyLoader();
+
+                // Allow this assembly loader to load any dll in assemblyFolderPath.
+                foreach (var dll in Directory.EnumerateFiles(assemblyFolderPath, "*.dll"))
+                {
+                    try
+                    {
+                        // Check if the file is a valid .NET assembly.
+                        AssemblyName.GetAssemblyName(dll);
+                    }
+                    catch
+                    {
+                        // The file is not a valid .NET assembly, skip it.
+                        continue;
+                    }
+
+                    analyzerAssemblyLoader.AddDependencyLocation(dll);
+                }
+
+                extension = new CustomMessageHandlerExtension(analyzerAssemblyLoader);
+                _extensions[assemblyFolderPath] = extension;
+            }
+        }
+
+        // AssemblyLoadLockObject is only used to avoid multiple calls from the same extensions to load the same assembly concurrently
+        // resulting in the constructors of the same handlers being called more than once.
+        // All other concurrent operations, including modifying extension.Assemblies are protected by _lockObject.
+        lock (extension.AssemblyLoadLockObject)
```

Comment:
do we need a separate lock for this?

alterantively, can this move into code *inside* 'extension' which locks internally, instead of exposing this lock outwards?